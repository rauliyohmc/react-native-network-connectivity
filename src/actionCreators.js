/* @flow */

import actionTypes from './actionTypes';
import type { FluxAction } from './types';

type EnqueuedAction = FluxAction | Function;

export const connectionChange = (isConnected: boolean): FluxAction => ({
  type: actionTypes.CONNECTION_CHANGE,
  payload: isConnected
});

export const fetchOfflineMode = (action: EnqueuedAction): FluxAction => {
  if (typeof action === 'object') {
    return {
      type: actionTypes.FETCH_OFFLINE_MODE,
      payload: {
        prevAction: {
          type: action.type,
          payload: action.payload
        }
      },
      meta: action.meta
    };
  }
  // Thunk
  return {
    type: actionTypes.FETCH_OFFLINE_MODE,
    payload: {
      prevThunk: action
    },
    meta: action.meta
  };
};

export const removeActionFromQueue = (action: EnqueuedAction): FluxAction => ({
  type: actionTypes.REMOVE_FROM_ACTION_QUEUE,
  payload: action
});

export const dismissActionsFromQueue = (actionTrigger: string): FluxAction => ({
  type: actionTypes.DISMISS_ACTIONS_FROM_QUEUE,
  payload: actionTrigger
});
