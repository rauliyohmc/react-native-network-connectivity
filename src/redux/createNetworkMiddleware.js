/* @flow */

import { find, get } from 'lodash';
import {
  fetchOfflineMode,
  removeActionFromQueue,
  dismissActionsFromQueue,
} from './actionCreators';
import type { NetworkState } from '../types';
import networkActionTypes from './actionTypes';
import wait from '../utils/wait';

type MiddlewareAPI<S> = {
  dispatch: (action: any) => void,
  getState(): S,
};

type State = {
  network: NetworkState,
};

type Arguments = {|
  regexActionType: RegExp,
  actionTypes: Array<string>,
  queueReleaseThrottle: number,
|};

function validateParams(regexActionType, actionTypes) {
  if ({}.toString.call(regexActionType) !== '[object RegExp]')
    throw new Error('You should pass a regex as regexActionType param');

  if ({}.toString.call(actionTypes) !== '[object Array]')
    throw new Error('You should pass an array as actionTypes param');
}

function findActionToBeDismissed(action, actionQueue) {
  return find(actionQueue, (a: *) => {
    const actionsToDismiss = get(a, 'meta.dismiss', []);
    return actionsToDismiss.includes(action.type);
  });
}

function isObjectAndShouldBeIntercepted(action, regexActionType, actionTypes) {
  return (
    typeof action === 'object' &&
    (regexActionType.test(action.type) || actionTypes.includes(action.type))
  );
}

function isThunkAndShouldBeIntercepted(action) {
  return typeof action === 'function' && action.interceptInOffline === true;
}

function checkIfActionShouldBeIntercepted(
  action,
  regexActionType,
  actionTypes,
) {
  return (
    isObjectAndShouldBeIntercepted(action, regexActionType, actionTypes) ||
    isThunkAndShouldBeIntercepted(action)
  );
}

function didComeBackOnline(action, wasConnected) {
  return (
    action.type === networkActionTypes.CONNECTION_CHANGE &&
    !wasConnected &&
    action.payload === true
  );
}

export const createReleaseQueue = (getState, next, delay) => async queue => {
  // eslint-disable-next-line
  for (const action of queue) {
    const { isConnected } = getState().network;
    if (isConnected) {
      next(removeActionFromQueue(action));
      next(action);
      // eslint-disable-next-line
      await wait(delay);
    } else {
      break;
    }
  }
};

function createNetworkMiddleware({
  regexActionType = /FETCH.*REQUEST/,
  actionTypes = [],
  queueReleaseThrottle = 50,
}: Arguments = {}) {
  return ({ getState }: MiddlewareAPI<State>) => (
    next: (action: any) => void,
  ) => (action: any) => {
    const { isConnected, actionQueue } = getState().network;
    const releaseQueue = createReleaseQueue(
      getState,
      next,
      queueReleaseThrottle,
    );
    validateParams(regexActionType, actionTypes);

    const shouldInterceptAction = checkIfActionShouldBeIntercepted(
      action,
      regexActionType,
      actionTypes,
    );

    if (shouldInterceptAction && isConnected === false) {
      // Offline, preventing the original action from being dispatched.
      // Dispatching an internal action instead.
      return next(fetchOfflineMode(action));
    }

    const isBackOnline = didComeBackOnline(action, isConnected);
    if (isBackOnline) {
      // Dispatching queued actions in order of arrival (if we have any)
      next(action);
      return releaseQueue(actionQueue);
    }

    // Checking if we have a dismissal case
    const isAnyActionToBeDismissed = findActionToBeDismissed(
      action,
      actionQueue,
    );
    if (isAnyActionToBeDismissed && !isConnected) {
      next(dismissActionsFromQueue(action.type));
    }

    // Proxy the original action to the next middleware on the chain or final dispatch
    return next(action);
  };
}

export default createNetworkMiddleware;
