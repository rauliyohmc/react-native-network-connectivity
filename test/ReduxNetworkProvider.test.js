import React from 'react';
import { View } from 'react-native';
import {
  shallow as rnShallow,
  render as rnRender,
} from 'react-native-testing-library';
import { shallow } from 'enzyme';
import {
  ReduxNetworkProvider,
  mapStateToProps,
} from '../src/components/ReduxNetworkProvider';
import { connectionChange } from '../src/redux/actionCreators';

const dispatch = jest.fn();

const props = {
  dispatch,
  isConnected: false,
  actionQueue: [],
};

const getProps = (overrides = {}) => ({ ...props, ...overrides });

describe('ReduxNetworkProvider', () => {
  afterEach(() => {
    dispatch.mockClear();
  });

  describe('render', () => {
    it('has the correct structure', () => {
      const { output } = rnShallow(
        <ReduxNetworkProvider>
          <View />
        </ReduxNetworkProvider>,
      );
      expect(output).toMatchSnapshot();
    });

    it('renders the children correctly', () => {
      const { getByTestId } = rnRender(
        <ReduxNetworkProvider>
          <View title="foo" testID="childrenView" />
        </ReduxNetworkProvider>,
      );
      const viewChild = getByTestId('childrenView');
      expect(viewChild.props.title).toBe('foo');
    });
  });

  describe('handleConnectivityChange', () => {
    it(`dispatches a CONNECTION_CHANGE action with the new connection`, () => {
      const wrapper = shallow(<ReduxNetworkProvider {...props} />);
      wrapper.instance().handleConnectivityChange(true);
      expect(props.dispatch).toHaveBeenCalledWith(connectionChange(true));
      expect(props.dispatch).toHaveBeenCalledTimes(1);
    });

    it(`does NOT dispatch a CONNECTION_CHANGE action if the connection
    did not change`, () => {
      const wrapper = shallow(
        <ReduxNetworkProvider {...getProps({ isConnected: true })} />,
      );
      wrapper.instance().handleConnectivityChange(true);
      expect(props.dispatch).not.toHaveBeenCalled();
    });
  });
});

describe('mapStateToProps', () => {
  it('maps isConnected and actionQueue state to props', () => {
    const networkState = {
      isConnected: false,
    };
    const state = {
      network: networkState,
    };
    expect(mapStateToProps(state)).toEqual(networkState);
  });
});
