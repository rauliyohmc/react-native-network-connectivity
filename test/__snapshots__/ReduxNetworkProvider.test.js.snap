// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReduxNetworkProvider render has the correct structure 1`] = `
<NetworkConnectivity
  httpMethod="HEAD"
  onConnectivityChange={[Function]}
  pingInBackground={false}
  pingInterval={0}
  pingOnlyIfOffline={false}
  pingServerUrl="https://www.google.com/"
  pingTimeout={10000}
  shouldPing={true}
>
  [Function]
</NetworkConnectivity>
`;
